### ISSUE TITLE
When is a value "Stable" - implementing stability logic for IOCs/blocks/Pvs  

### ISSUE THREAD

**Issue by FreddieAkeroyd** [ISISComputingGroup/IBEX#3149]
This is a starter ticket for a discussion on implementing "stability" of values in IBEX, such as "is my temperature stable". The general idea is that depending on PID settings a value may oscillate about its setpoint for a while before settling down. Applying "run control" would stop data collection when out of range, but this is best used for values that might drift slowly out of range, if values are oscillating you would be continually doing a stop/start and collecting something that may not be what you want.  

what is stable - typically this would mean "within some specified range for an unbroken timespan of ..." but there may be other appropriate calculations that could be applied to a range of recent values.

stability could be applied to PVs at the IOC level, but applying it to blocks (like we do for run control) may be most flexible. There is nothing to stop us adding it to select IOCs too.

Muons currently use a client side stability criteria when they initially ramp the temperature, but after that they collect regardless. So you may want to calculate stability without actually doing something additional, so a script could "waitfor stable" before doing a begin. Other poeple may want a future "unstable" to pause collection like run control does, or have the IOC do an initial stability calculation followed by normal run control when things settle down (this could be accomplished via a waitfor_stable plus normal run_control). Though using stability for stop/start for a whole run would work, if values are slowly drifting in and out rather than oscillating using a time based stability would discard more data that was likely OK.

can we use run control limits as stability limts, or do we need another set of limits?

stability calculations and other control loogic should be able to be handled by additional generated records loaded by the existing run control IOC.

**Comment by kjwoodsISIS** [comment]
Thanks, Freddie, for kicking this off.

Some things that occur to me:
1. How do we define stability?  We might need more than one definition, to cope with different types of situation.  And we should write down these definitions - to avoid future misunderstandings.  For example, stability could mean
   1. a variable x, which varies in time, is strictly constrained to remain within two limits: a <= x(t) <= b, but otherwise we don't care how x behaves.
   1. a variable y, which varies in time, converges on a defined value and then remains close to that defined value: y(t) -> k +/- eps (where eps is some small value)
   1. a variable z, which varies in time, is desired to remain within two limits: a <= z(t) <= b, but if perturbed outside the desired range will return (within some defined period of time).
   1. There is a lot of similarity between these definitions, so ...
1. We should create an abstracted definition of stability, so that we can apply it to different situations.
1. Many devices have built-in stability mechanisms.  How are we (and indeed scientists) taking advantage of these?  
   1. What happens if we layer our own stability control code on top of the existing stability mechanisms?  It seems to me that in some cases, at least, there is a risk that one stability mechanism might interfere with the other (e.g. if both are not correctly set up, one could prevent the other from reaching a stable state - which would, of course, defeat the purpose of both mechanisms).  So we probably need to create our own guidance here.

**Comment by ChrisM-S** [comment]
Yes, this is somewhere we could make IBEX very strong.  I would generalize the definition of "stable" to be
the boolean `stable(t) =    a(t) <= z(t)  <= b(t) `

**Comment by ChrisM-S** [comment]
Also, given that the scientists have declared an "interest" in taking data where possible (probably indicated by having started a run), is there anything else you would do other than pause/stop taking data when the critical value left it's trajectory/envelope?.

I agree, it would probably be a good idea to specifically avoid any risk of doubling up on on control as @Kevin-Woods-Tessella mentioned in (3) above, i.e. not get into any more decisions or actions other than pause/resume.

An envelope specification would cover the startup case when a temperature was reached (the limit functions `a(t)` and `b(t)` could also take temperature or other experiment parameters into account (`t` was specifically for time!).

A trajectory and limit envelope might also be useful in motion control to detect when something is starting not to move smoothly e.g. an oscillating collimator.

**Comment by FreddieAkeroyd** [comment]
I think most cases are contained in a "range" + "time in range" with a time of zero equating more to current run control. Specifying the range as a delta or percentage is more flexible than specifying limits that would need adjusting each temperature change, but this is likely to be done in a script so might not be an issue? Another possible criteria might be "standard deviation over a particular range", but we need to talk to scientists as what is meaningful.

**Comment by FreddieAkeroyd** [comment]
At the moment there is no proposal to have our data collecting stability criteria do anything active to the device, it would just suspend data collection. Device stability should be done in the IOC for the device itself?

**Comment by ChrisM-S** [comment]
Agreed, that might be an argument for setting up the record logic on the blocks because they can be more arbitrary and can change (whereas the device may have very different criteria for stability than the experiment).

**Comment by kjwoodsISIS** [comment]
I can imagine situations where the range is not symmetric about the desired stable value - say -5%, but +10%.

**Comment by ChrisM-S** [comment]
Yes, this is the sort of thing (or maybe even logarithm related) I had a discussion some while back with someone about pressure cells where there is an expected gradual fall off in pressure (I think due to natural deformation or absorption) after pressurisation.  This would be fine for data taking, but should a seal fail and the pressure drop right off the Mega Pascal range, the sample would be compromised and the data would be invalidated.  As this is a sort of reverse example, this has always stuck in my mind.  If we were able to offer a couple of python limit functions for the more awkward cases like this, the simple situations would be catered for by simply returning constant limits.

**Comment by kjwoodsISIS** [comment]
I am not sure I fully understand this.
1. What is the significance of "unstable" here?  Can they just not make measurements regardless of the "stability" of the block?  Are they asking for some new capability here?  (if so, I don't understand what it is).
1. This is just run control: "when some condition is true, collect data", although we will need to extend the range of conditions that we support (see next comment).

**Comment by kjwoodsISIS** [comment]
To my mind we seem to have two aspects to this discussion:
1. one is run-control and whether we might need to extend our run-control capability to support other types of condition
2. the notion of stability of a block (or PV) and how we might define stability 

Run control is, fundamentally, "collect data while <some condition> is true".  From the above discussion, I see four ways of defining the condition.
1.  Condition 1: collect data while a block, x, lies with a defined range [a, b].  In other words, while (a <= x <= b).
1.  Condition 2: collect data while a block, x, lies with a defined range [a, b], but x is allowed to stray out of range for a limited period.
In other words, while ((a <= x <= b) OR (( x < a OR x > b) AND ((t - t0) < Tmax)))), where t is the current time, t0 is the time x first strayed out of range and t-max is a maximum period (defined by the user).
There needs to be some logic to manage the resetting of t0 (after x comes back within range, but before it next goes out of range).
1.  Condition 3: collect data while a block, x, lies with a defined range [l, u] relative to a desired value, m.  l and u are the lower & upper limits relative to m
In other words, while (m - l <= x <= m + u).  l and u might be expressed as percentages.
1.  Condition 4: collect data while a block, x, lies with a defined range [l, u] relative to a desired value, m, , but x is allowed to stray out of range for a limited period.
In other words, while ((m - l <= x <= m + u) OR (( x < m - l OR x > m + b) AND ((t - t0) < Tmax)))).

I think this covers most cases.  You could argue that 1 and 2 are special cases of 3 and 4, but users might find it more convenient to specify absolute limits in some situations and relative limits in others.  Statistical measures are covered by case 3 (and 4) (for example: standard deviations about a mean - m could be the mean (or some other measure), l and u could be some multiple of the standard deviation).

All 4 conditions are, arguably, definitions of stability.  There might be other definitions of stability.  However, so long as the definition of stability always results in a statement that resolves to TRUE or FALSE, we can use that definition as a run-control condition.

We might also want to use definitions of stability in other circumstances (e.g. the variable in question might not be a block, it might be for control of a device, rather than control of data collection)).  So it would be useful to have definitions of stability that are abstracted from the specific circumstances, so that we can apply them to any new set of circumstances.

**Comment by FreddieAkeroyd** [comment]
The SANDALS use case is the same as muons currently use, enabled by a client side stability calculation. They want to do a run while it is "getting to temperature" which is their "unstable", and then a run at constant temperature.

cset(new temp)
begin()
while (true)
    cget(temp)
    if (temp not in range) 
           stable_start_time = time.now()
    if time.now() - stable_start_start > stable_time 
          break;
end()
set temp  low and high limits for run control
begin()
waitfor( uamps)
end()

In their case a "waitfor_stable" genie command with some choices of stability calculation function would work. It is slightly different from run control as currently run control pauses/resumes collection based on some criteria whereas here they want to start/end. However it would be possible to do it with a sort of run control if you configured the action as well as the formula, but you would still need to have the script wait for something so you might have waitfor_setup etc. and then you might wonder if waitfor_stable is better. It might be useful fro some instruments to have stability instead of our current run control, but one the temperature has reasonable stabilised this might lead to additional good data being thrown away. So the notion of "stability" is dependent on which phase of the experiment is in progress, and they may want to collect this data (sandals) or just know when it is finished (muons)

**Comment by John-Holt-Tessella** [comment]
+1 CRISP

**Comment by John-Holt-Tessella** [comment]
See #4230 for more requirements

**Comment by John-Holt-Tessella** [comment]
There was a tech talk question 25th April 2019 entitles "database or new record for blocking setpoint/readback" which suggested using the https://github.com/epics-modules/lua module

**Comment by ThomasLohnert** [comment]
Is this still needed or is this sufficiently addressed by https://github.com/ISISComputingGroup/EPICS-utilities/pull/14?
